mlrun/api/main.py

```diff
@@ -224,6 +226,105 @@ async def move_api_to_online():
+    if config.log_collector.mode == mlrun.api.schemas.LogsCollectorMode.legacy:
+        logger.info(
+            "Using legacy logs collection method, skipping logs collection periodic function",
+            mode=config.log_collector.mode,
+        )
+        return
+    logger.info(
+        "Starting logs collection periodic function",
+        mode=config.log_collector.mode,
+        interval=config.log_collector.interval,
+    )
+    run_function_periodically(
+        config.log_collector.interval,
+        _collect_runs_logs.__name__,
+        False,
+        _collect_runs_logs,
+    )
+
+
+start_logs_limit = asyncio.Semaphore(config.log_collector.concurrent_start_logs_workers)
+
+
+async def _collect_runs_logs():
+    db_session = create_session()
+    try:
+        # list all the runs in the system which we didn't request logs collection for yet
+        runs = await fastapi.concurrency.run_in_threadpool(
+            get_db().list_distinct_runs_uids,
+            db_session,
+            project="*",
+            requested_logs=False,
+            only_uids=False,
+        )
+        # each result contains either run_uid or None
+        # if it's None it means something went wrong and we should skip it
+        # if it's run_uid it means we requested logs collection for it and we should update it's requested_logs field
+        results = await asyncio.gather(*[_start_log_for_run(run) for run in runs])
+        runs_to_update_requested_logs = [result for result in results if result]
+
+        if len(runs_to_update_requested_logs) > 0:
+            # update the runs to indicate that we have requested log collection for them
+            await fastapi.concurrency.run_in_threadpool(
+                get_db().update_runs_requested_logs,
+                db_session,
+                uids=runs_to_update_requested_logs,
+            )
+
+    finally:
+        close_session(db_session)
+
+
+async def _start_log_for_run(run: dict) -> typing.Union[str, None]:
+    """
+    Starts log collection for a specific run
+    :param run: run object
+    :return: the run_uid of the run if log collection was started, None otherwise
+    """
+    # using semaphore to limit the number of concurrent log collection requests
+    # this is to prevent opening too many connections to many connections
+    async with start_logs_limit:
+        logs_collector_client = (
+            mlrun.api.utils.clients.log_collector.LogCollectorClient()
+        )
+        run_kind = run.get("metadata", {}).get("labels", {}).get("kind", None)
+        project_name = run.get("metadata", {}).get("project", None)
+        run_uid = run.get("metadata", {}).get("uid", None)
+        # empty kind means it's a local run, the log collector doesn't support it
+        if not run_kind:
+            # we mark the run as requested logs collection so we won't iterate over it again
+            return run_uid
+        try:
+            runtime_handler = get_runtime_handler(run_kind)
```

Comment:
blocking io, see how it calls `resolve_mpijob_crd_version` which goes to k8s in a blocking way